rs/ledger_suite/icp/ledger/src/main.rs

```diff
@@ -745,11 +748,59 @@ fn main() {
     })
 }
 
+// We use 8MiB buffer
+const BUFFER_SIZE: usize = 8388608;
+
 fn post_upgrade(args: Option<LedgerCanisterPayload>) {
     let start = dfn_core::api::performance_counter(0);
-    let mut stable_reader = stable::StableReader::new();
+
+    // In order to read the first bytes we need to use ic_cdk.
+    // dfn_core assumes the first 4 bytes store stable memory length
+    // and return bytes starting from the 5th byte.
+    let mut magic_bytes_reader = ic_cdk::api::stable::StableReader::default();
+    const MAGIC_BYTES: &[u8; 3] = b"MGR";
+    let mut first_bytes = [0u8; 3];
+    let memory_manager_found = match magic_bytes_reader.read_exact(&mut first_bytes) {
+        Ok(_) => first_bytes == *MAGIC_BYTES,
```

Comment:
Could it be that the dfn_core also write "MGR" as the first three bytes because the binary encoding of the stable memory length would accidentally happen to start with "MGR"?

Reply:
Good point! In addition to the magic bytes "MGR", the memory manager also writes [the layout version](https://github.com/dfinity/stable-structures/blob/817a163e0f8a5a2afaad0469e9434398ac6bf7d3/src/memory_manager.rs#L97), which is a `1u8`. So the first four bytes would be `[0x77, 0x71, 0x82, 0x01]`, and this would correspond to a stable memory size of 25325943 bytes (`StableWriter` [writes the length](https://sourcegraph.com/github.com/dfinity/ic@4eca90d6ecc9027e743b61f0a526b7b59ba5663b/-/blob/rs/rust_canisters/dfn_core/src/stable.rs?L105-109) `u32` as `le_bytes`). Looking at [the heap memory usages of the ledger canisters](https://grafana.mainnet.dfinity.network/explore?orgId=1&left=%7B%22datasource%22:%22P119963096DF10E67%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P119963096DF10E67%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22ledger_total_memory_bytes%20%25%202%20%21%3D%200%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%22now-90d%22,%22to%22:%22now%22%7D%7D), they are all using an even amount of bytes. Based only on this, and without having looked into the details on how the serialization is done (if it could end up being an odd number of bytes), I would say that it is very unlikely that we would run into a situation where dfn_core would end up writing the same four bytes as the memory manager, WDYT?

Reply:
> it could be a coincidence that the sizes are all even

Nevertheless, maybe an acceptable risk to take.

Reply:
I did some further local testing where I wrote `b"MGR" + bytes([1]` to the first bytes of raw stable memory in the pre-upgrade without memory manager, such that in the V1 post-upgrade the code would detect the memory manager. In this case, when thereafter trying to deserialize the state, expecting that it was written using the memory manager, the canister trapped. A subsequent downgrade back to the V0 mainnet version was successful, so even if it happened that V0 wrote `b"MGR" + bytes([1]` as the stable memory size, it is very likely that the subsequent memory manager initialization would fail. This would still be an issue for SNS ICRC ledgers, in the sense that they would require a fixed version to be able to further upgrade, but at least there wouldn't be any data loss/corruption. This is all anyway extremely unlikely to happen, so as you said, it may be an acceptable risk to take.

Reply:
> A subsequent downgrade back to the V0 mainnet version was successful

I guess this is because the actual size of the stable memory that was read was less than 
```
>>> int.from_bytes(b"MGR" + bytes([1]), "little")
22169421
```
(otherwise `dfn_core` would likely fail)

Reply:
on the other hand, in practice, the actual size would not be overriden by `b"MGR" + bytes([1])`, so the downgrade would actually be possible